common/src/main/java/com/microsoft/identity/common/internal/dto/IdToken.java

```diff
@@ -27,4 +27,22 @@ public String getRealm() {
     public void setRealm(final String realm) {
         mRealm = realm;
     }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
for my learning, i want to understand these 2 functions

Reply:
[This is the best summary I could find](http://www.baeldung.com/java-hashcode) from Googling around.

In our case, I needed to `@Override` these functions because [some of the unit tests use `equals()`](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/fe63948102ebb4f296a77d66e364c461a1b735d5/common/src/androidTest/java/com/microsoft/identity/common/AccountCredentialCacheTest.java#L81) to verify that the `Account` or `Credential` restored from the cache is _essentially_ the same object as the one which was written to it, despite the fact that their memory addresses are different.

I've overridden `hashCode()` because, in classes where `equals()` has been overridden, `hashCode()` **must** also be overridden ([docs](https://docs.oracle.com/javase/tutorial/java/IandI/objectclass.html): See section 'The `equals()` method').

>By definition, if two objects are equal, their hash code must also be equal. If you override the `equals()` method, you change the way two objects are equated and `Object`'s implementation of `hashCode()` is no longer valid. Therefore, if you override the `equals()` method, you must also override the `hashCode()` method as well.

Regarding the use of `31` as the multiplier: [there seems to be some hand-waving going on about the nuts and bolts of the optimization](https://stackoverflow.com/a/3613764/741827) but in practical terms, we can say that using a multiplier with numerous divisors results in more hash collisions so a prime number represents the 'best choice'. I used it because it is the default implementation provided by Android Studio 🤷‍♂️